snapshot/lib/assets/SnapshotHelper.swift

```diff
@@ -181,7 +181,7 @@ open class Snapshot: NSObject {
 
                 let path = screenshotsDir.appendingPathComponent("\(simulator)-\(name).png")
                 #if swift(<5.0)
```

Comment:
> But at some point this was broken with Swift 5.0, another if clause should be added to know when it was added the throws statement. Probably in the latest release with Swift 5.7

Not sure what you mean @minuscorp 🙈 could you clarify/rephrase?

Should we do something like

```
if swift <4.2
    try
else if swift <5.0
    don't try, just call it
else 
    no change
```

?

Reply:
What is the Swift version you're targetting @amcalgates ? We should prob test this in <4.2, <5.0 (e.g. 4.2), and latest (5.7)

Reply:
That's it, we support back to Swift 4.2 until now so it's probable that if you compile in Swift 5.5 this new code breaks because the throws API was not introduced yet

Reply:
I'm concerned about why the code has changed, but I think it was a later fix to adapt to all code in Swift 5.X and just fallback to the Swift 4 version otherwise, but I don't know if I made the change afterwards. 🤔

Reply:
>I'm concerned about why the code has changed

🤔 You mean why this swift file is broken again?

The changes made here https://github.com/fastlane/fastlane/pull/18930 were only to the example app, which's not part of fastlane swift generation source code, it's just a sample app. This PR changes the actual template file used to generate the swift files when running the init command - so effectively it was still broken in that PR, and it seemed fixed because the steps to reproduce the issue (and its fix) weren't right (running the Example app aren't the right reproducibility steps).

Looking forward to a better approach, I think we should generate the Example app files dynamically after each deploy, just like our `docs` repo and other swift files, based on the actual source files. Perhaps even move them to a `/Generated` folder to make this more clear 😬 (nice-to-have, out of the scope of the current PR, but I think that'd be the best way to keep a single source of truth) - What do you think?

Reply:
It seems like a good idea, but I don't know if it's worth the effort, we would be pushed to use XcodeGen or another tool in order to create the example on every push doesn't it?

Reply:
Not necessarily I think, we could have the the Xcode projects with multiple setups (such as Swift 3, 4, 5), and then fastlane would just generate (update, really) the swift files that swift-related init actions would generate

Reply:
The major issue is to update the run script for each runner, maintain different code because different ABI, I think it can be harder than it seems, or its just my Sunday head

Reply:
Maybe I'm not following, the ABI differences don't need to change from the current implementation we have - I'm just suggesting we should generate the code from our templates, populate them in Example projects that can be compiled so we can verify that the generated code actually compiles. And then do this for 3 difference major Swift API versions OR (my preference) deprecate Swift 3 and 4 and support only 5+ 😜 But we can probably get away with 2: 4.2 and 5.0+

Reply:
I don't know about the project, but the package uses SPM 4.2+ right now. Swift 3 is so unstable🤦‍♂️

Reply:
Anyway, let's firstly fix the issue with Swift 5.0 compilation and then we can address to make the CI compile with different Swift versions the example project.

Reply:
> Which Swift version are you targeting @amcalgates ? We should prob test this in <4.2, <5.0 (e.g. 4.2), and latest (5.7)

In build settings it says I'm targeting Swift 5.

Reply:
There are only two build systems, Swift 4 and Swift 5. Generically, you're compiling with the Swift version embedded in your Xcode version, 5.7 probably. But build system != Swift version compiling the code.

Reply:
> Which Swift version are you targeting @amcalgates ? We should prob test this in <4.2, <5.0 (e.g. 4.2), and latest (5.7)

Ah, yeah sorry about that. 95% of my job is still objc so I'm a little fuzzy on some of the Swift specifics. My answer certainly seemed wrong, but I didn't think to look outside of build settings. Here's the actual answer:

```
% swift -version
swift-driver version: 1.45.2 Apple Swift version 5.6.1 (swiftlang-5.6.0.323.66 clang-1316.0.20.12)
```

Reply:
@minuscorp re:

>That's it, we support back to Swift 4.2 until now so it's probable that if you compile in Swift 5.5 this new code breaks because the throws API was not introduced yet

Since @amcalgates is on Swift 5.6.1 already, don't you think this code is ok as is? 🤔

Reply:
I need to investigate which API were shipped with which version of the stdlib which matches some Swift version... I'll have to install a couple of Xcode versions to check this

Reply:
Edit 3: yup haha exactly my point in the beginning of this thread 😆  fixing the wrong spot haha
Edit 2 is a good question though 🤔

Reply:
@amcalgates you should be checking the Swift version of the Xcode project for fastlane, not your main project's. Just double checking 🙇

Reply:
Apologies for the delay, Covid finally got me after all this time.

@rogerluan - you'll need to be more specific, unfortunately. I gave the output of `swift -version` from the command line (which doesn't take an Xcode project as an argument). Previously, I said "Swift 5" because that's what Build Settings says, but that is not the desired answer, either. That basically exhausts all the options I found in the top google hit [here](https://stackoverflow.com/questions/30790188/how-do-i-see-which-version-of-swift-im-using).

Reply:
@amcalgates no worries!
If I remember correctly, Fastlane Swift provides you with a .xcodeproj of its own for you to work with and implement your Fastfile.swift, Deliver.swift, etc — all of fastlane files. You then also have your project's .xcodeproj, e.g. "YourCompany.xcodeproj" (or ".xcworkspace"). We need to your fastlane .xcodeproj's Swift version (found in the Build Settings), and not your company's xcodeproj. Up there it wasn't clear whether the swift version was read from your project's build settings or from fastlane's project's build settings, so I just wanted to make sure we're on the same page here 🙏 because, as @minuscorp said [here](https://github.com/fastlane/fastlane/pull/20689#discussion_r982223319):

>Edit 2: Why does it fail to compile if you're using a Swift >5.0 version? It shouldn't go through that compile branch.

Makes us think that you're actually on Swift 4.2 and not on Swift 5.

Thank you!

Reply:
> Edit 2: Why does it fail to compile if you're using a Swift >5.0 version? It shouldn't go through that compile branch.

I could not make sense of this, either. I'm not a swift expert, but in objc-land that's a compiler flag and shouldn't be triggered.

Where it's failing _is_ in my project, though. `fastlane snapshot init` copies `SnapshotHelper.swift` to "YourCompany.xcodeproj"'s folder. Then, when I try to run UITests, I get the same error as in https://github.com/fastlane/fastlane/issues/18792.

I'm not using Fastlane Swift, so I'm not being provided an xcodeproj by fastlane. I'm using vanilla fastlane + snapshot with a fastfile.

All this forced me to re-evaluate my project settings. I'm actually programmatically creating the UITests using https://github.com/tuist/XcodeProj. I'm not specifying Swift version for my UITests target, so that's why I'm hitting the error - my UITests target seems to default to Swift 4. 🤦‍♂️

So, no cause for alarm - compiler flags work fine in Swift 🙃

I assume my fix is still relevant, since there are two copies of `SnapshotHelper.swift`. It's none of my business, but I agree with @rogerluan above that having a single copy of this file is probably a good long-term goal. Perhaps as part of the example users should acutally run `fastlane snapshot init` and anything else that copies files over? Maybe having the examples in their own repo would help?

Reply:
>fastlane snapshot init copies SnapshotHelper.swift to "YourCompany.xcodeproj"'s folder. Then, when I try to run UITests, I get the same error as in https://github.com/fastlane/fastlane/issues/18792.
>I'm not using Fastlane Swift, so I'm not being provided an xcodeproj by fastlane. I'm using vanilla fastlane + snapshot with a fastfile.

Ah gotcha! I had this all wrong, thanks for explaining it to me! 🙏 

>my UITests target seems to default to Swift 4. 🤦‍♂️

This clears all questions we had 😁 

Thank you again for guiding us through this! I'll create an issue to track the improvements we discovered in this discussion 🙌